# Creating the Initial Slice


In [9]:
output_table = fablib.list_sites()

In [1]:
from fabrictestbed_extensions.fablib.fablib import FablibManager as fablib_manager
fablib = fablib_manager()

In [2]:
slice = fablib.new_slice(name="INTProject")

In [3]:
#End host 1
site1='SRI'   #SRI
#Switches
site2='UCSD'  #UCSD
site3='STAR'  #STAR
site4='MAX'  #ATLA
#End host 2
site5='RUTG'  #WASH

print (f'The selected sites are {site1}, {site2}, {site3}, {site4}, {site5}') 

The selected sites are SRI, UCSD, STAR, MAX, RUTG


## Create Nodes


In [4]:
server1 = slice.add_node(name="server1", 
                      site=site1, 
                      cores=4, 
                      ram=8, 
                      disk=20, 
                      image='default_ubuntu_20')

server2 = slice.add_node(name="server2", 
                      site=site5, 
                      cores=4, 
                      ram=8, 
                      disk=20, 
                      image='default_ubuntu_20')
monitor1 = slice.add_node(name="monitor1", 
                      site=site2, 
                      cores=4, 
                      ram=8, 
                      disk=20, 
                      image='default_ubuntu_20')
monitor2 = slice.add_node(name="monitor2", 
                      site=site4, 
                      cores=4, 
                      ram=8, 
                      disk=20, 
                      image='default_ubuntu_20')
server1_iface = server1.add_component(model='NIC_Basic').get_interfaces()[0]
server2_iface = server2.add_component(model='NIC_Basic').get_interfaces()[0]
monitor1_iface = monitor1.add_component(model='NIC_Basic').get_interfaces()[0]
monitor2_iface = monitor2.add_component(model='NIC_Basic').get_interfaces()[0]

## P4 Switches Nodes

In [5]:
switch1 = slice.add_node(name="switch1", 
                      site=site2, 
                      cores=16, 
                      ram=16, 
                      disk=40, 
                      image='default_ubuntu_20')
switch2 = slice.add_node(name="switch2", 
                      site=site3, 
                      cores=16, 
                      ram=16, 
                      disk=40, 
                      image='default_ubuntu_20')
switch3 = slice.add_node(name="switch3", 
                      site=site4, 
                      cores=16, 
                      ram=16, 
                      disk=40, 
                      image='default_ubuntu_20')
switch_iface1 = switch1.add_component(model='NIC_Basic', name='net1_nic').get_interfaces()[0]
switch_iface2 = switch1.add_component(model='NIC_Basic', name='net2_nic').get_interfaces()[0]
switch_iface3 = switch2.add_component(model='NIC_Basic', name='net1_nic').get_interfaces()[0]
switch_iface4 = switch2.add_component(model='NIC_Basic', name='net2_nic').get_interfaces()[0]
switch_iface5 = switch3.add_component(model='NIC_Basic', name='net1_nic').get_interfaces()[0]
switch_iface6 = switch3.add_component(model='NIC_Basic', name='net2_nic').get_interfaces()[0]

switch_sink_iface1 = switch1.add_component(model='NIC_Basic', name='net3_nic').get_interfaces()[0]
switch_sink_iface2 = switch3.add_component(model='NIC_Basic', name='net3_nic').get_interfaces()[0]

## Connecting Sites

In [6]:
net1 = slice.add_l2network(name='net1', interfaces=[server1_iface, switch_iface1])
net2 = slice.add_l2network(name='net2', interfaces=[switch_iface2, switch_iface3])
net3 = slice.add_l2network(name='net3', interfaces=[switch_iface4, switch_iface5])
net4 = slice.add_l2network(name='net4', interfaces=[switch_iface6, server2_iface])

net5 = slice.add_l2network(name='net5', interfaces=[switch_sink_iface1, monitor1_iface])
net6 = slice.add_l2network(name='net6', interfaces=[switch_sink_iface2, monitor2_iface])


## Submit Slice

In [7]:
slice.submit();


Retry: 11, Time: 314 sec



Time to print interfaces 359 seconds


## Configure Sites

In [8]:
#Install bmv2 and net tools on switches
for x in range (1,4):
    print(x)
    switch = slice.get_node(name="switch" + str(x))     
    switch.upload_file('scripts/install_bmv2.sh', 'install_bmv2.sh')
    stdout, stderr = switch.execute(f'chmod +x install_bmv2.sh &&  ./install_bmv2.sh',quiet=True)
    stdout, stderr = switch.execute(f'sudo apt-get install -y net-tools', quiet=True)

1
2
3


In [9]:
#Net tools and scapy on servers
server1 = slice.get_node(name="server1")
server2 = slice.get_node(name="server2")

monitor1 = slice.get_node(name="monitor1")
monitor2 = slice.get_node(name="monitor2")

stdout, stderr = server1.execute(f'sudo apt-get install -y net-tools', quiet=True)
stdout, stderr = server2.execute(f'sudo apt-get install -y net-tools', quiet=True)

stdout, stderr = monitor1.execute(f'sudo apt-get install -y net-tools', quiet=True)
stdout, stderr = monitor2.execute(f'sudo apt-get install -y net-tools', quiet=True)

stdout, stderr = server1.execute(f'sudo apt-get update && sudo apt-get install -y python3-scapy', quiet=True)
stdout, stderr = server2.execute(f'sudo apt-get update && sudo apt-get install -y python3-scapy', quiet=True)

stdout, stderr = monitor1.execute(f'sudo apt-get update && sudo apt-get install -y python3-scapy', quiet=True)
stdout, stderr = monitor2.execute(f'sudo apt-get update && sudo apt-get install -y python3-scapy', quiet=True)


In [10]:
#Interfaces
node1_iface = server1.get_interface(network_name='net1') 
server1_iface_name = node1_iface.get_device_name()
print(f'server1_iface: {server1_iface_name}')

node2_iface = server2.get_interface(network_name='net4') 
server2_iface_name = node2_iface.get_device_name()
print(f'server2_iface: {server2_iface_name}')


monitor1_iface = monitor1.get_interface(network_name='net5') 
monitor1_iface_name = monitor1_iface.get_device_name()
print(f'monitor1_iface: {monitor1_iface_name}')

monitor2_iface = monitor2.get_interface(network_name='net6') 
monitor2_iface_name = monitor2_iface.get_device_name()
print(f'monitor2_iface: {monitor2_iface_name}')

switch1 = slice.get_node(name="switch1")
switch2 = slice.get_node(name="switch2")
switch3 = slice.get_node(name="switch3")


switch1_iface1 = switch1.get_interface(network_name='net1') 
switch1_iface1_name = switch1_iface1.get_device_name()
print(f'switch_iface1: {switch1_iface1_name}')
stdout, stderr = switch.execute(f'sudo ip link set {switch1_iface1_name} mtu 2000')

switch1_iface2 = switch1.get_interface(network_name='net2') 
switch1_iface2_name = switch1_iface2.get_device_name()
print(f'switch_iface2: {switch1_iface2_name}')
stdout, stderr = switch1.execute(f'sudo ip link set {switch1_iface2_name} mtu 2000')


switch_sink_iface1 = switch1.get_interface(network_name='net5') 
switch_sink_iface1_name = switch_sink_iface1.get_device_name()
print(f'switch_sink_iface1: {switch_sink_iface1_name}')
stdout, stderr = switch1.execute(f'sudo ip link set {switch_sink_iface1_name} mtu 2000')

switch2_iface1 = switch2.get_interface(network_name='net2') 
switch2_iface1_name = switch2_iface1.get_device_name()
print(f'switch_iface3: {switch2_iface1_name}')
stdout, stderr = switch2.execute(f'sudo ip link set {switch2_iface1_name} mtu 2000')

switch2_iface2 = switch2.get_interface(network_name='net3') 
switch2_iface2_name = switch2_iface2.get_device_name()
print(f'switch_iface4: {switch2_iface2_name}')
stdout, stderr = switch2.execute(f'sudo ip link set {switch2_iface2_name} mtu 2000')


switch3_iface1 = switch3.get_interface(network_name='net3') 
switch3_iface1_name = switch3_iface1.get_device_name()
print(f'switch_iface5: {switch3_iface1_name}')
stdout, stderr = switch3.execute(f'sudo ip link set {switch3_iface1_name} mtu 2000')

switch3_iface2 = switch3.get_interface(network_name='net4') 
switch3_iface2_name = switch3_iface2.get_device_name()
print(f'switch_iface6: {switch3_iface2_name}')
stdout, stderr = switch3.execute(f'sudo ip link set {switch3_iface2_name} mtu 2000')

switch_sink_iface2 = switch3.get_interface(network_name='net6') 
switch_sink_iface2_name = switch_sink_iface2.get_device_name()
print(f'switch_sink_iface2: {switch_sink_iface2_name}')
stdout, stderr = switch3.execute(f'sudo ip link set {switch_sink_iface2_name} mtu 2000')


server1_iface: enp7s0
server2_iface: enp7s0
monitor1_iface: enp7s0
monitor2_iface: enp7s0
switch_iface1: enp7s0
switch_iface2: enp8s0
switch_sink_iface1: enp9s0
switch_iface3: enp8s0
switch_iface4: enp7s0
switch_iface5: enp8s0
switch_iface6: enp9s0
switch_sink_iface2: enp7s0


In [11]:
#Turn interfaces up
stdout, stderr = server1.execute(f'sudo ip link set dev {server1_iface_name} up', quiet=True)
stdout, stderr = server2.execute(f'sudo ip link set dev {server2_iface_name} up', quiet=True)

stdout, stderr = monitor1.execute(f'sudo ip link set dev {monitor1_iface_name} up', quiet=True)
stdout, stderr = monitor2.execute(f'sudo ip link set dev {monitor2_iface_name} up', quiet=True)

stdout, stderr = switch1.execute(f'sudo ip link set dev {switch1_iface1_name} up', quiet=True)
stdout, stderr = switch1.execute(f'sudo ip link set dev {switch1_iface2_name} up', quiet=True)
stdout, stderr = switch1.execute(f'sudo ip link set dev {switch_sink_iface1_name} up', quiet=True)
stdout, stderr = switch2.execute(f'sudo ip link set dev {switch2_iface1_name} up', quiet=True)
stdout, stderr = switch2.execute(f'sudo ip link set dev {switch2_iface2_name} up', quiet=True)
stdout, stderr = switch3.execute(f'sudo ip link set dev {switch3_iface1_name} up', quiet=True)
stdout, stderr = switch3.execute(f'sudo ip link set dev {switch3_iface2_name} up', quiet=True)
stdout, stderr = switch3.execute(f'sudo ip link set dev {switch_sink_iface2_name} up', quiet=True)

In [12]:
#Hardcode MAC addresses
server1_iface_MAC = '00:00:00:00:00:01'
switch1_iface1_MAC = '00:00:00:00:00:02'
switch1_iface2_MAC = '00:00:00:00:00:03'
switch2_iface1_MAC = '00:00:00:00:00:04'
switch2_iface2_MAC = '00:00:00:00:00:05'
switch3_iface1_MAC = '00:00:00:00:00:06'
switch3_iface2_MAC = '00:00:00:00:00:07'
server2_iface_MAC = '00:00:00:00:00:08'

switch_sink_iface1_MAC = '00:00:00:00:00:09'
monitor1_iface_MAC = '00:00:00:00:00:0A'

switch_sink_iface2_MAC = '00:00:00:00:00:0B'
monitor2_iface_MAC = '00:00:00:00:00:0C'


In [13]:
#IP addresses
server1_switch1_subnet = "192.168.1.0/24"
server1_ip = '192.168.1.1/24'
switch1_ip1 = '192.168.1.2/24'

stdout, stderr = server1.execute(f'sudo ifconfig {server1_iface_name} {server1_ip}')
stdout, stderr = switch1.execute(f'sudo ifconfig {switch1_iface1_name} {switch1_ip1}')

stdout, stderr = server1.execute(f'sudo ifconfig {server1_iface_name} hw ether {server1_iface_MAC}')
stdout, stderr = switch1.execute(f'sudo ifconfig {switch1_iface1_name} hw ether {switch1_iface1_MAC}')

switch1_switch2_subnet = "192.168.2.0/24"
switch1_ip2 = '192.168.2.1/24'
switch2_ip1 = '192.168.2.2/24'

stdout, stderr = switch1.execute(f'sudo ifconfig {switch1_iface2_name} {switch1_ip2}')
stdout, stderr = switch2.execute(f'sudo ifconfig {switch2_iface1_name} {switch2_ip1}')

stdout, stderr = switch1.execute(f'sudo ifconfig {switch1_iface2_name} hw ether {switch1_iface2_MAC}')
stdout, stderr = switch2.execute(f'sudo ifconfig {switch2_iface1_name} hw ether {switch2_iface1_MAC}')

switch2_switch3_subnet = "192.168.3.0/24"
switch2_ip2 = '192.168.3.1/24'
switch3_ip1 = '192.168.3.2/24'

stdout, stderr = switch2.execute(f'sudo ifconfig {switch2_iface2_name} {switch2_ip2}')
stdout, stderr = switch3.execute(f'sudo ifconfig {switch3_iface1_name} {switch3_ip1}')

stdout, stderr = switch2.execute(f'sudo ifconfig {switch2_iface2_name} hw ether {switch2_iface2_MAC}')
stdout, stderr = switch3.execute(f'sudo ifconfig {switch3_iface1_name} hw ether {switch3_iface1_MAC}')

switch3_server2_subnet = "192.168.4.0/24"
switch3_ip2 = '192.168.4.1/24'
server2_ip = '192.168.4.2/24'

stdout, stderr = switch3.execute(f'sudo ifconfig {switch3_iface2_name} {switch3_ip2}')
stdout, stderr = server2.execute(f'sudo ifconfig {server2_iface_name} {server2_ip}')

stdout, stderr = switch3.execute(f'sudo ifconfig {switch3_iface2_name} hw ether {switch3_iface2_MAC}')
stdout, stderr = server2.execute(f'sudo ifconfig {server2_iface_name} hw ether {server2_iface_MAC}')

switch1_monitor1_subnet = "192.168.5.0/24"
switch1_ip3 = '192.168.5.1/24'
monitor1_ip = '192.168.5.2/24'

stdout, stderr = switch1.execute(f'sudo ifconfig {switch_sink_iface1_name} {switch1_ip3}')
stdout, stderr = monitor1.execute(f'sudo ifconfig {monitor1_iface_name} {monitor1_ip}')

stdout, stderr = switch1.execute(f'sudo ifconfig {switch_sink_iface1_name} hw ether {switch_sink_iface1_MAC}')
stdout, stderr = monitor1.execute(f'sudo ifconfig {monitor1_iface_name} hw ether {monitor1_iface_MAC}')

switch3_monitor2_subnet = "192.168.6.0/24"
switch3_ip3 = '192.168.6.1/24'
monitor2_ip = '192.168.6.2/24'

stdout, stderr = switch3.execute(f'sudo ifconfig {switch_sink_iface2_name} {switch3_ip3}')
stdout, stderr = monitor2.execute(f'sudo ifconfig {monitor2_iface_name} {monitor2_ip}')

stdout, stderr = switch3.execute(f'sudo ifconfig {switch_sink_iface2_name} hw ether {switch_sink_iface2_MAC}')
stdout, stderr = monitor2.execute(f'sudo ifconfig {monitor2_iface_name} hw ether {monitor2_iface_MAC}')



In [14]:
#Enable forwards
command = 'sudo sysctl -w net.ipv4.ip_forward=1' 
stdout, stderr = switch1.execute(command, quiet=True)
stdout, stderr = switch2.execute(command, quiet=True)
stdout, stderr = switch3.execute(command, quiet=True)

In [15]:
#Delete Routing

stdout, stderr = switch1.execute(f'sudo ip route del {server1_switch1_subnet}', quiet=True)
stdout, stderr = switch1.execute(f'sudo ip route del {switch3_server2_subnet}', quiet=True)

stdout, stderr = switch2.execute(f'sudo ip route del {server1_switch1_subnet}', quiet=True)
stdout, stderr = switch2.execute(f'sudo ip route del {switch3_server2_subnet}', quiet=True)

stdout, stderr = switch3.execute(f'sudo ip route del {server1_switch1_subnet}', quiet=True)
stdout, stderr = switch3.execute(f'sudo ip route del {switch3_server2_subnet}', quiet=True)

In [16]:
#Configure Routing
gw1 = switch1_ip1.split('/')[0]
gw2 = switch3_ip2.split('/')[0]
print(gw1, gw2)
stdout, stderr = server1.execute(f'sudo ip route add {switch3_server2_subnet} via {gw1}')
stdout, stderr = server2.execute(f'sudo ip route add {server1_switch1_subnet} via {gw2}')
stdout, stderr = server1.execute(f'sudo arp -s {gw1} {switch1_iface1_MAC}')
stdout, stderr = server2.execute(f'sudo arp -s {gw2} {switch3_iface2_MAC}')
gw1 = switch1_ip3.split('/')[0]
gw2 = switch3_ip3.split('/')[0]
stdout, stderr = monitor1.execute(f'sudo arp -s {gw1} {switch_sink_iface1_MAC}')
stdout, stderr = monitor2.execute(f'sudo arp -s {gw2} {switch_sink_iface2_MAC}')

"""
gw1 = server1_ip.split('/')[0]
gw2 = switch2_ip1.split('/')[0]
stdout, stderr = switch1.execute(f'sudo arp -s {gw1} {server1_iface_MAC}')
stdout, stderr = switch1.execute(f'sudo arp -s {gw2} {switch2_iface1_MAC}')

gw1 = switch2_ip1.split('/')[0]
gw2 = switch2_ip2.split('/')[0]
stdout, stderr = switch1.execute(f'sudo ip route add {switch3_server2_subnet} via {gw1}')
stdout, stderr = switch3.execute(f'sudo ip route add {server1_switch1_subnet} via {gw2}')

gw1 = switch3_ip1.split('/')[0]
gw2 = switch1_ip2.split('/')[0]
stdout, stderr = switch2.execute(f'sudo ip route add {switch3_server2_subnet} via {gw1}')
stdout, stderr = switch2.execute(f'sudo ip route add {server1_switch1_subnet} via {gw2}')

"""

192.168.1.2 192.168.4.1


"\ngw1 = server1_ip.split('/')[0]\ngw2 = switch2_ip1.split('/')[0]\nstdout, stderr = switch1.execute(f'sudo arp -s {gw1} {server1_iface_MAC}')\nstdout, stderr = switch1.execute(f'sudo arp -s {gw2} {switch2_iface1_MAC}')\n\ngw1 = switch2_ip1.split('/')[0]\ngw2 = switch2_ip2.split('/')[0]\nstdout, stderr = switch1.execute(f'sudo ip route add {switch3_server2_subnet} via {gw1}')\nstdout, stderr = switch3.execute(f'sudo ip route add {server1_switch1_subnet} via {gw2}')\n\ngw1 = switch3_ip1.split('/')[0]\ngw2 = switch1_ip2.split('/')[0]\nstdout, stderr = switch2.execute(f'sudo ip route add {switch3_server2_subnet} via {gw1}')\nstdout, stderr = switch2.execute(f'sudo ip route add {server1_switch1_subnet} via {gw2}')\n\n"

In [17]:
from fabrictestbed_extensions.fablib.fablib import FablibManager as fablib_manager
fablib = fablib_manager()
slice = fablib.get_slice(name="INTProject")
server1 = slice.get_node(name="server1")
server2 = slice.get_node(name="server2")
switch1 = slice.get_node(name='switch1')        
switch2 = slice.get_node(name='switch2')        
switch3 = slice.get_node(name='switch3')        
monitor1 = slice.get_node(name='monitor1')        
monitor2 = slice.get_node(name='monitor2')        



In [18]:
ssh = server1.get_ssh_command()
print(ssh)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@192.5.67.82


In [19]:
ssh = switch1.get_ssh_command()
command = (
    f"simple_switch -i 0@{switch1_iface1_name} -i 1@{switch1_iface2_name} "
    f"-i 2@{switch_sink_iface1_name} basic.json --log-console &"
)
print(ssh)
print(command)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@132.249.252.164
simple_switch -i 0@enp7s0 -i 1@enp8s0 -i 2@enp9s0 basic.json --log-console &


In [20]:
ssh = switch2.get_ssh_command()
command = (
    f"simple_switch -i 0@{switch2_iface1_name} -i 1@{switch2_iface2_name} "
    f"basic.json --log-console"
)
print(ssh)
print(command)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@2001:400:a100:3030:f816:3eff:fea7:8621
simple_switch -i 0@enp8s0 -i 1@enp7s0 basic.json --log-console


In [21]:
ssh = switch3.get_ssh_command()
command = (
    f"simple_switch -i 0@{switch3_iface1_name} -i 1@{switch3_iface2_name} "
    f"-i 2@{switch_sink_iface2_name} basic.json --log-console &"
)
print(ssh)
print(command)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@2001:468:c00:ffc4:f816:3eff:fedb:5f2f
simple_switch -i 0@enp8s0 -i 1@enp9s0 -i 2@enp7s0 basic.json --log-console &


In [22]:
ssh = server2.get_ssh_command()
print(ssh)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@2620:0:d61:4101:f816:3eff:feeb:a487


In [23]:
ssh = monitor1.get_ssh_command()
print(ssh)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@132.249.252.184


In [24]:
ssh = monitor2.get_ssh_command()
print(ssh)

ssh -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@2001:468:c00:ffc4:f816:3eff:fea0:dcd9


sudo su

p4c INTProject/p4/basic.p4

p4c INTProject/p4Switch1/basic.p4

p4c INTProject/p4Switch2/basic.p4

p4c INTProject/p4Switch3/basic.p4

simple_switch -i 0@enp7s0 -i 1@enp8s0 basic.json --log-console

simple_switch -i 0@enp8s0 -i 1@enp7s0 basic.json --log-console &

simple_switch -i 0@enp8s0 -i 1@enp9s0 basic.json --log-console &

simple_switch -i 0@enp9s0 -i 1@enp8s0 basic.json --log-console &

simple_switch -i 0@enp7s0 -i 1@enp9s0 basic.json --log-console &

simple_switch -i 0@enp9s0 -i 1@enp7s0 basic.json --log-console &

simple_switch -i 0@enp7s0 -i 1@enp9s0 -i 2@enp8s0 basic.json --log-console &

simple_switch -i 0@enp9s0 -i 1@enp7s0 -i 2@enp8s0 basic.json --log-console &

simple_switch -i 0@enp9s0 -i 1@enp8s0 -i 2@enp7s0 basic.json --log-console &

simple_switch -i 0@enp7s0 -i 1@enp8s0 -i 2@enp9s0 basic.json --log-console &

simple_switch -i 0@enp8s0 -i 1@enp9s0 -i 2@enp7s0 basic.json --log-console &


pkill -t pts/1

simple_switch_CLI

pkill simple_switch

mirroring_add 1 2

In [35]:
#Upload p4 programs
switch1 = slice.get_node(name='switch1')        
switch1.upload_directory('p4Switch1', '/home/ubuntu/INTProject')
switch1.upload_directory('p4', '/home/ubuntu/INTProject')

switch2 = slice.get_node(name='switch2')        
switch2.upload_directory('p4Switch2', '/home/ubuntu/INTProject')
switch2.upload_directory('p4', '/home/ubuntu/INTProject')

switch3 = slice.get_node(name='switch3')        
switch3.upload_directory('p4', '/home/ubuntu/INTProject')
switch3.upload_directory('p4Switch3', '/home/ubuntu/INTProject')

'success'

In [36]:
#Upload rules 
#Make sure programs are running!
switch1.upload_file('rules/switch1Rules.sh', 'rules.sh')
stdout, stderr = switch1.execute('chmod +x rules.sh && ./rules.sh')

switch2.upload_file('rules/switch2Rules.sh', 'rules.sh')
stdout, stderr = switch2.execute('chmod +x rules.sh && ./rules.sh')

switch3.upload_file('rules/switch3Rules.sh', 'rules.sh')
stdout, stderr = switch3.execute('chmod +x rules.sh && ./rules.sh')

Obtaining JSON from switch...
Done
Control utility for runtime P4 table manipulation
RuntimeCmd: Adding entry to lpm match table MyIngress.ipv4_lpm
match key:           LPM-c0:a8:01:00/24
action:              MyIngress.forward
runtime data:        00:00:00:00:00:01	00:00
Entry has been added with handle 0
RuntimeCmd: 
Obtaining JSON from switch...
Done
Control utility for runtime P4 table manipulation
RuntimeCmd: Adding entry to exact match table MyIngress.ipv4_exact
match key:           EXACT-c0:a8:01:01
action:              MyIngress.forward
runtime data:        00:00:00:00:00:01	00:00
Entry has been added with handle 0
RuntimeCmd: 
Obtaining JSON from switch...
Done
Control utility for runtime P4 table manipulation
RuntimeCmd: Adding entry to lpm match table MyIngress.ipv4_lpm
match key:           LPM-c0:a8:04:00/24
action:              MyIngress.forward
runtime data:        00:00:00:00:00:04	00:01
Entry has been added with handle 1
RuntimeCmd: 
Obtaining JSON from switch...
Done
Co

In [29]:
monitor1 = slice.get_node(name='monitor1')        
monitor1.upload_file('monitor.py', 'monitor.py')
monitor2 = slice.get_node(name='monitor2')        
monitor2.upload_file('monitor.py', 'monitor.py')

<SFTPAttributes: [ size=724 uid=1000 gid=1000 mode=0o100664 atime=1733556714 mtime=1733556714 ]>

In [153]:
import time
server1.upload_file('p4/send.py', 'send.py')
#server1.execute(f'sudo python3 send.py {server1_iface} 192.168.1.10 192.168.2.10 HelloWorld')
stdout, stderr = server1.execute(f'sudo python3 send.py enp7s0 192.168.1.1 192.168.4.2 HelloWorld')
#    time.sleep(0.5)

sending on interface enp7s0 to 192.168.4.2
###[ Ethernet ]### 
  dst       = 00:00:00:00:00:02
  src       = 00:00:00:00:00:01
  type      = IPv4
###[ IP ]### 
     version   = 4
     ihl       = 5
     tos       = 0x0
     len       = 51
     id        = 1
     flags     = 
     frag      = 0
     ttl       = 64
     proto     = tcp
     chksum    = 0xf470
     src       = 192.168.1.1
     dst       = 192.168.4.2
     \options   \
###[ TCP ]### 
        sport     = 52253
        dport     = 1234
        seq       = 0
        ack       = 0
        dataofs   = 5
        reserved  = 0
        flags     = S
        window    = 8192
        chksum    = 0x99a
        urgptr    = 0
        options   = []
###[ Raw ]### 
           load      = '192.168.1.1'



sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.lo.disable_ipv6=1

tshark -i enp7s0 -w captures/newCapture.pcap

chmod 644 /home/ubuntu/captures/newCapture.pcap

scp -r -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@[2620:0:d61:4101:f816:3eff:fec0:4850]:~/captures ~/work/Project/captures

scp -r -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@[2001:468:c00:ffc4:f816:3eff:fe44:88d]:~/telemetry_data.csv ~/work/Project/telemetry

scp -r -i /home/fabric/work/fabric_config/slice_key -F /home/fabric/work/fabric_config/ssh_config ubuntu@132.249.252.164:~/captures ~/work/Project/captures

132.249.252.157

iperf3 -c 192.168.4.2 -u -b 100G -t 120 -P 10 -l 1000
iperf3 -c 192.168.4.2 -u -b 10G -t 30 -l 1000
iperf3 -c 192.168.4.2 -u -b 1M -t 30 -l 1000

In [180]:
from datetime import datetime
from datetime import timezone
from datetime import timedelta

#Set end host to now plus 1 day
end_date = (datetime.now(timezone.utc) + timedelta(days=8)).strftime("%Y-%m-%d %H:%M:%S %z")

try:
    slice = fablib.get_slice(name="INTProject")

    slice.renew(end_date)
except Exception as e:
    print(f"Exception: {e}")


Retry: 0, Time: 90 sec



Time to print interfaces 135 seconds


In [ ]:
server1 = slice.get_node(name="server1")
server2 = slice.get_node(name="server2")
stdout, stderr = server1.execute(f'sudo apt-get install -y net-tools', quiet=True)
stdout, stderr = server2.execute(f'sudo apt-get install -y net-tools', quiet=True)


In [28]:
slice = fablib.get_slice(name="INTProject")
slice.delete()

Plan
-Append test metadata at each switch
-Remove metadata when next hop is destination

-Set up Monitor
-Send removed metadata to monitor